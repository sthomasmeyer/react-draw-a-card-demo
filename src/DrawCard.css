.DrawCard {
  height: 400px;
  width: 775px;
  outline: 5px solid crimson;
  background-color: navy;
  margin: 0 auto;
  padding: 11px;
  align-items: center;
}

.DrawCard h1 {
  color: ghostwhite;
  text-align: center;
  margin-bottom: 15px;
  margin-top: 5px;
}

.DrawCard div {
  position: relative;
  height: 300px;
  display: flex;
  justify-content: center;
}

.DrawCard button {
  display: inline-block;
  background-color: ghostwhite;
  color: navy;
  padding: 8px;
}

.DrawCard button:hover {
  cursor: pointer;
}

.DrawCard .AutoDrawBtn {
  float: right;
}

.DrawCard .DrawCardBtn {
  float: left;
}
