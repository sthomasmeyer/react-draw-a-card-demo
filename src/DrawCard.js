import React, { useState, useEffect } from 'react';
import './DrawCard.css';
import axios from 'axios';
import Card from './Card';

const DrawCard = () => {
  const [deckId, setDeckId] = useState([]);
  const [card, setCard] = useState([]);
  const [autoDraw, setAutoDraw] = useState(false);
  const [exhaustedDeck, setExhaustedDeck] = useState(false);

  useEffect(() => {
    console.log(`React's [useEffect()] Hook is working.`);
    console.log(card);
    async function getNewDeck() {
      const res = await axios.get(
        `http://deckofcardsapi.com/api/deck/new/shuffle/?deck_count=1`
      );
      console.log(`New Deck ID: ${res.data.deck_id}`);
      setDeckId(res.data.deck_id);
    }

    getNewDeck();
  }, []);

  async function drawCardOnClick() {
    console.log(`The [drawCardOnClick()] function has been executed.`);
    const res = await axios.get(
      `http://deckofcardsapi.com/api/deck/${deckId}/draw/?count=1`
    );

    if (res.data.success) {
      setCard(res.data.cards[0]);
    } else {
      setExhaustedDeck(true);
      setCard([]);
    }
  }

  const activateAutoDraw = () => setAutoDraw(true);
  const deactivateAutoDraw = () => setAutoDraw(false);

  useEffect(() => {
    if (autoDraw === true && !exhaustedDeck) {
      const intervalId = setInterval(() => drawCardOnClick(), 1000);
      return () => clearInterval(intervalId);
    }
  }, [autoDraw]);

  async function reShuffleDeck() {
    const res = await axios.get(
      `http://deckofcardsapi.com/api/deck/${deckId}/shuffle/`
    );
    console.log(
      `The deck has been re-shuffled successfully: ${res.data.success}`
    );
    res.data.success
      ? setExhaustedDeck(false)
      : console.error(
          `Failed 'GET' request. The deck has *not* been re-shuffled.`
        );
  }

  return (
    <div className='DrawCard'>
      <h1>Draw a Card</h1>
      <div>
        <Card key={card.code} id={card.code} image={card.image} />
      </div>

      {!exhaustedDeck ? (
        <button onClick={drawCardOnClick} id='DrawCardBtn'>
          Draw a Card
        </button>
      ) : null}

      {!exhaustedDeck && autoDraw ? (
        <button onClick={deactivateAutoDraw} className='AutoDrawBtn'>
          Pause Auto Draw
        </button>
      ) : null}

      {!exhaustedDeck && !autoDraw ? (
        <button onClick={activateAutoDraw} className='AutoDrawBtn'>
          Auto Draw Cards
        </button>
      ) : null}

      {exhaustedDeck ? (
        <button onClick={reShuffleDeck} id='ReShuffleBtn'>
          Re-shuffle the Deck
        </button>
      ) : null}
    </div>
  );
};

export default DrawCard;
