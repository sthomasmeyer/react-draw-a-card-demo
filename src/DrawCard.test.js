import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, screen, fireEvent } from '@testing-library/react';
import DrawCard from './DrawCard';

// Smoke test:
it('renders without crashing', () => {
  render(<DrawCard />);
});

it('matches snapshot', () => {
  const { asFragment } = render(<DrawCard />);
  expect(asFragment()).toMatchSnapshot();
});

it('draws a card effectively on-click', () => {
  render(<DrawCard />);

  const drawCardBtn = screen.getByRole('button', {
    name: /Draw a Card/i
  });
  fireEvent.click(drawCardBtn);

  const playingCard = screen.queryByAltText('playing card');
  expect(playingCard).toBeInTheDocument();
});
