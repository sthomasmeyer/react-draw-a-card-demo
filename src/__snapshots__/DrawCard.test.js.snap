// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="DrawCard"
  >
    <h1>
      Draw a Card
    </h1>
    <div>
      <div
        class="Card"
      >
        <img
          alt="playing card"
        />
      </div>
    </div>
    <button
      id="DrawCardBtn"
    >
      Draw a Card
    </button>
    <button
      class="AutoDrawBtn"
    >
      Auto Draw Cards
    </button>
  </div>
</DocumentFragment>
`;
